precompile/allowlist/allowlist.go

```diff
@@ -181,11 +187,24 @@ func CreateAllowListPrecompile(precompileAddr common.Address) contract.StatefulP
 }
 
 func CreateAllowListFunctions(precompileAddr common.Address) []*contract.StatefulPrecompileFunction {
-	setAdmin := contract.NewStatefulPrecompileFunction(setAdminSignature, createAllowListRoleSetter(precompileAddr, AdminRole))
-	setManager := contract.NewStatefulPrecompileFunctionWithActivator(setManagerSignature, createAllowListRoleSetter(precompileAddr, ManagerRole), contract.IsDUpgradeActivated)
-	setEnabled := contract.NewStatefulPrecompileFunction(setEnabledSignature, createAllowListRoleSetter(precompileAddr, EnabledRole))
-	setNone := contract.NewStatefulPrecompileFunction(setNoneSignature, createAllowListRoleSetter(precompileAddr, NoRole))
-	read := contract.NewStatefulPrecompileFunction(readAllowListSignature, createReadAllowList(precompileAddr))
-
-	return []*contract.StatefulPrecompileFunction{setAdmin, setManager, setEnabled, setNone, read}
+	var functions []*contract.StatefulPrecompileFunction
+
+	for name, method := range AllowListABI.Methods {
+		var fn *contract.StatefulPrecompileFunction
+		if name == "readAllowList" {
+			fn = contract.NewStatefulPrecompileFunction(method.ID, createReadAllowList(precompileAddr))
+		} else if adminFnName, _ := AdminRole.GetSetterFunctionName(); name == adminFnName {
+			fn = contract.NewStatefulPrecompileFunction(method.ID, createAllowListRoleSetter(precompileAddr, AdminRole))
+		} else if enabledFnName, _ := EnabledRole.GetSetterFunctionName(); name == enabledFnName {
+			fn = contract.NewStatefulPrecompileFunction(method.ID, createAllowListRoleSetter(precompileAddr, EnabledRole))
+		} else if noRoleFnName, _ := NoRole.GetSetterFunctionName(); name == noRoleFnName {
+			fn = contract.NewStatefulPrecompileFunction(method.ID, createAllowListRoleSetter(precompileAddr, NoRole))
+		} else if managerFnName, _ := ManagerRole.GetSetterFunctionName(); name == managerFnName {
+			fn = contract.NewStatefulPrecompileFunctionWithActivator(method.ID, createAllowListRoleSetter(precompileAddr, ManagerRole), contract.IsDUpgradeActivated)
+		} else {
+			panic(fmt.Sprintf("unexpected method name: %s", name))
+		}
+		functions = append(functions, fn)
+	}
+	return functions
```

Comment:
what's the reason for this change?

Reply:
We introduced ABI to allow list. We also return error from `GetSetterFunctionName`.